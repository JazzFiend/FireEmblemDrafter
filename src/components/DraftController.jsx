import React, { Component } from 'react';
import PropTypes from 'prop-types';
import _ from 'lodash';

import Pick from './Pick';
import RandomNumberGenerator from '../logic/RandomNumberGenerator';
import RandomElementSelector from '../logic/RandomElementSelector';
import RosterOptions from '../logic/helpers/RosterOptions';
import Draft from '../logic/Draft';
import TestRandomNumberGenerator from '../logic/__mocks__/TestRandomNumberGenerator';

class DraftController extends Component {
  static displayTeam(team) {
    let teamDisplayFriendly = '';
    _.forEach(team, (value) => {
      teamDisplayFriendly = `${teamDisplayFriendly + value}\n`;
    });
    return teamDisplayFriendly;
  }

  static createRandomizer(randomizePicks) {
    if (randomizePicks) {
      return new RandomElementSelector(RandomNumberGenerator);
    }
    TestRandomNumberGenerator.clearState();
    return new RandomElementSelector(TestRandomNumberGenerator);
  }

  constructor(props) {
    super(props);
    this.state = {
      pick: Array(6),
      team: Array(0),
      draft: undefined,
    };
  }

  showPick() {
    const { draftInProgress, handleDraftProgress } = this.props;
    const { pick, draft } = this.state;
    return (
      <div>
        {
          draftInProgress
          && (
          <Pick
            pick={pick}
            onClick={(index) => {
              this.handlePickClick(index);
              handleDraftProgress(!(draft.isDraftFinished()));
            }}
          />
          )
        }
      </div>
    );
  }

  handlePickClick(buttonIndex) {
    const { pick, draft } = this.state;
    let newPick;

    draft.select(pick[buttonIndex]);
    if (!draft.isDraftFinished()) {
      newPick = draft.generateNextPick();
    }

    this.setState({
      pick: newPick,
      team: draft.getCurrentTeam(),
    });
  }

  startDraft() {
    const {
      roster,
      restrictedCharacters,
      requiredCharacters,
      teamSize,
      randomizePicks,
    } = this.props;
    const rosterOptions = new RosterOptions(roster, restrictedCharacters, requiredCharacters);
    const randomizer = DraftController.createRandomizer(randomizePicks);
    const draft = new Draft(rosterOptions, teamSize, randomizer);
    const pick = draft.generateNextPick();

    this.setState({
      draft,
      team: draft.getCurrentTeam(),
      pick,
    });
  }

  render() {
    const { draftInProgress, handleDraftProgress } = this.props;
    const { team } = this.state;
    return (
      <div>
        <button
          type="button"
          className="start-draft-button"
          data-testid="start-draft-button"
          onClick={() => {
            this.startDraft();
            handleDraftProgress(true);
          }}
        >
          {draftInProgress ? 'Restart Draft!' : 'Start Draft!'}
        </button>
        {this.showPick()}
        <div data-testid="team-list">
          {DraftController.displayTeam(team)}
        </div>
      </div>
    );
  }
}

DraftController.propTypes = {
  draftInProgress: PropTypes.bool.isRequired,
  handleDraftProgress: PropTypes.func,
  roster: PropTypes.arrayOf(PropTypes.string),
  restrictedCharacters: PropTypes.arrayOf(PropTypes.string),
  requiredCharacters: PropTypes.arrayOf(PropTypes.string),
  teamSize: PropTypes.number.isRequired,
  randomizePicks: PropTypes.bool,
};

DraftController.defaultProps = {
  handleDraftProgress: () => {},
  roster: [],
  restrictedCharacters: [],
  requiredCharacters: [],
  randomizePicks: true,
};

export default DraftController;
