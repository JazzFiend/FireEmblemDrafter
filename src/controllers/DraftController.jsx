import React, { Component } from 'react';
import PropTypes from 'prop-types';

import RandomNumberGenerator from '../model/RandomNumberGenerator';
import RandomElementSelector from '../model/RandomElementSelector';
import RosterOptions from '../model/helpers/RosterOptions';
import Draft from '../model/Draft';
import TestRandomNumberGenerator from '../model/__mocks__/TestRandomNumberGenerator';
import DraftView from '../view/DraftView';

class DraftController extends Component {
  static createRandomizer(randomizePicks) {
    if (randomizePicks) {
      return new RandomElementSelector(RandomNumberGenerator);
    }
    TestRandomNumberGenerator.clearState();
    return new RandomElementSelector(TestRandomNumberGenerator);
  }

  constructor(props) {
    super(props);
    this.state = {
      pick: Array(6),
      draft: undefined,
    };
  }

  selectPick(buttonIndex) {
    const { pick, draft } = this.state;
    let newPick;

    draft.select(pick[buttonIndex]);
    if (!draft.isDraftFinished()) {
      newPick = draft.generateNextPick();
    }

    this.setState({
      pick: newPick,
    });
  }

  extractTeam() {
    const { draft } = this.state;
    if (draft) {
      return draft.getCurrentTeam();
    }
    return [];
  }

  startDraft() {
    const {
      roster,
      restrictedCharacters,
      requiredCharacters,
      teamSize,
      randomizePicks,
    } = this.props;
    const rosterOptions = new RosterOptions(roster, restrictedCharacters, requiredCharacters);
    const randomizer = DraftController.createRandomizer(randomizePicks);
    const draft = new Draft(rosterOptions, teamSize, randomizer);
    const pick = draft.generateNextPick();

    this.setState({
      draft,
      pick,
    });
  }

  render() {
    const {
      draftInProgress,
      handleDraftProgress,
    } = this.props;

    const { draft, pick } = this.state;

    const handleStartDraft = () => {
      this.startDraft();
      handleDraftProgress(true);
    };

    const pickHandler = (index) => {
      this.selectPick(index);
      handleDraftProgress(!(draft.isDraftFinished()));
    };

    return (
      <DraftView
        draftInProgress={draftInProgress}
        handleStartDraft={handleStartDraft}
        handlePickClick={pickHandler}
        pick={pick}
        team={this.extractTeam()}
      />
    );
  }
}

DraftController.propTypes = {
  draftInProgress: PropTypes.bool.isRequired,
  handleDraftProgress: PropTypes.func,
  roster: PropTypes.arrayOf(PropTypes.string),
  restrictedCharacters: PropTypes.arrayOf(PropTypes.string),
  requiredCharacters: PropTypes.arrayOf(PropTypes.string),
  teamSize: PropTypes.number.isRequired,
  randomizePicks: PropTypes.bool,
};

DraftController.defaultProps = {
  handleDraftProgress: () => {},
  roster: [],
  restrictedCharacters: [],
  requiredCharacters: [],
  randomizePicks: true,
};

export default DraftController;
