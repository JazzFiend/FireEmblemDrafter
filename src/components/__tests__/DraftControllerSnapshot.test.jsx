import React from 'react';
import { render, cleanup } from '@testing-library/react';
import DraftController from '../DraftController';

jest.mock('../Pick', () => () => 'Pick');

afterEach(cleanup);

test('renders default Draft Controller when draft is in progress', () => {
  const { asFragment } = render(
    <DraftController
      draftInProgress={true}
      handleDraftProgress={jest.fn()}
      roster={['1', '2', '3']}
      restrictedCharacters={['1']}
      requiredCharacters={['2']}
      teamSize={1}
      randomizePicks={false}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});

test('renders default Draft Controller when draft hasn\'t started', () => {
  const { asFragment } = render(
    <DraftController
      draftInProgress={false}
      handleDraftProgress={jest.fn()}
      roster={['1', '2', '3']}
      restrictedCharacters={['1']}
      requiredCharacters={['2']}
      teamSize={1}
      randomizePicks={false}
    />,
  );
  expect(asFragment()).toMatchSnapshot();
});
