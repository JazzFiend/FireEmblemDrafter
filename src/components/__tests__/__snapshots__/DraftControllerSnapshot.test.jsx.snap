// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders default Draft Controller when draft hasn't started 1`] = `
<DocumentFragment>
  <div>
    <button
      class="start-draft-button"
      data-testid="start-draft-button"
      type="button"
    >
      Start Draft!
    </button>
    <div />
    <div
      data-testid="team-list"
    />
  </div>
</DocumentFragment>
`;

exports[`renders default Draft Controller when draft is in progress 1`] = `
<DocumentFragment>
  <div>
    <button
      class="start-draft-button"
      data-testid="start-draft-button"
      type="button"
    >
      Restart Draft!
    </button>
    <div>
      Pick
    </div>
    <div
      data-testid="team-list"
    />
  </div>
</DocumentFragment>
`;
